From e42e01ff1a741159bcb4974eea9285ee6b9185b8 Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Thu, 14 May 2026 22:59:00 +0200
Subject: [PATCH 04/48] ntdll: Use unwind_context values only if the caller is
 not EC in capture_context.

(cherry picked from commit 3433e08073e1f742fc8442532fe8d189b745641b)

CW-Bug-Id: #27308
---
 dlls/ntdll/signal_arm64ec.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/signal_arm64ec.c b/dlls/ntdll/signal_arm64ec.c
index 3a8576d3e1b..0b9cb5b8328 100644
--- a/dlls/ntdll/signal_arm64ec.c
+++ b/dlls/ntdll/signal_arm64ec.c
@@ -1430,7 +1430,8 @@ static void __attribute__((used)) capture_context( CONTEXT *context, UINT cpsr,
     /* unwind one level to get register values from caller function */
     unwind_context = *context;
     unwind_one_frame( &unwind_context );
-    memcpy( &context->Rax, &unwind_context.Rax, offsetof(CONTEXT,FltSave) - offsetof(CONTEXT,Rax) );
+    if (!RtlIsEcCode( unwind_context.Rip ))
+        memcpy( &context->Rax, &unwind_context.Rax, offsetof(CONTEXT,FltSave) - offsetof(CONTEXT,Rax) );
 }
 
 /***********************************************************************
-- 
2.55.0

